fix(race-conditon): fix data race condition when accessing `stderr` on pipe

mahadzaryab1 opened the pull request:
There was a race condition that existed in `Pipe.Exec`, which spins up a goroutine that access `Pipe.stderr` without a lock. This PR fixes that by protecting `Pipe.stderr` with `Pipe.mu`. This PR fixes #193. 

**Testing**
Created a simple unit test as follows: 
```go
func TestDataRaceStdErr(t *testing.T) {
	stdOut := new(bytes.Buffer)
	stdErr := new(bytes.Buffer)

	_, err := script.Exec("echo").WithStdout(stdOut).WithStderr(stdErr).Stdout()
	if err != nil {
		t.Fatal(err)
	}
}
```
Running `go test -race` before the changes in this PR outputted the same warning seen in the issue. Running it again with the changes in the PR cause the tests to pass.

mahadzaryab1 commented:
question: we currently do not run `go test` in the CI with the `-race` flag so I removed the regression test. Do we want to run the tests with the `-race` flag? If so, should I add the regression test back in?

bitfield commented:
If we can _have_ race conditions, we should definitely run CI tests with the race detector on!

mahadzaryab1 commented:
> If we can _have_ race conditions, we should definitely run CI tests with the race detector on!

Added the `-race` flag to `ci.yml` and added a regression test.

ccoVeille commented on script.go:
```diff
@@ -385,7 +385,8 @@ func (p *Pipe) Exec(cmdLine string) *Pipe {
 		cmd.Stdout = w
 		cmd.Stderr = w
-		if p.stderr != nil {
+		pipeStderr := p.getStderr()
+		if pipeStderr != nil {
 			cmd.Stderr = p.stderr
 		}
```
I have a problem with the logic here

getStderr returns pipeStdErr


```suggestion
		pipeStderr := p.getStderr()
		if pipeStderr != nil {
 			cmd.Stderr = pipeStderr
 		}
```

While the code is somehow similar, even if there could remain some races, because we are accessing the p.stderr directly, no ?

ccoVeille replied:
If not, the method and signature could be updated, because there is no need to return it, so,it's not a getter

mahadzaryab1 replied:
@ccoVeille this was an oversight! thanks for catching - fixed now

ccoVeille commented on script.go:
```diff
@@ -425,7 +426,8 @@ func (p *Pipe) ExecForEach(cmdLine string) *Pipe {
 			cmd.Stdout = w
 			cmd.Stderr = w
-			if p.stderr != nil {
+			pipeStderr := p.getStderr()
+			if pipeStderr != nil {
 				cmd.Stderr = p.stderr
 			}
```
```suggestion
			pipeStderr := p.getStderr()
			if pipeStderr != nil {
 				cmd.Stderr = pipeStderr
 			}
```

bitfield commented on script.go:
```diff
@@ -31,7 +31,7 @@ type Pipe struct {
 	httpClient     *http.Client
 
-	// because pipe stages are concurrent, protect 'err'
+	// because pipe stages are concurrent, protect 'err' and 'stderr'
```
Notice how the `err` field is listed below the `mu` field, in its own 'block', separated from the other fields by a blank line. This is a nice way to show, lexically, that the mutex is 'over' these fields. To maintain this, let's move the `stderr` field in underneath the mutex too.

Is there any reason not to include the `stdout` field too? I mean, if a race condition exists on `stderr`, doesn't it also exist on `stdout`?

mahadzaryab1 replied:
> Notice how the `err` field is listed below the `mu` field, in its own 'block', separated from the other fields by a blank line. This is a nice way to show, lexically, that the mutex is 'over' these fields. To maintain this, let's move the `stderr` field in underneath the mutex too.
addressed!
> Is there any reason not to include the `stdout` field too? I mean, if a race condition exists on `stderr`, doesn't it also exist on `stdout`?
The race condition exists on `stderr` because it is accessed inside a goroutine at https://github.com/bitfield/script/blob/master/script.go#L388-L390. The same is not true for `stdout` so I don't believe a race condition exists on `stdout`. We could still protect it with a mutex to match the behavior of `stdout` if we'd like however. Let me know what you think.

bitfield commented on script.go:
```diff
@@ -762,6 +764,30 @@ func (p *Pipe) SetError(err error) {
+	if p.mu == nil { // uninitialised pipe
+		return nil
+	}
+	p.mu.Lock()
```
```suggestion
	p.mu.RLock()
```

I think we only need a read lock here.

mahadzaryab1 replied:
@bitfield we can't do this since we're using a `sync.Mutex` instead of a `sync.RWMutex`

bitfield replied:
This needs a little more thought, doesn't it? If we're just _retrieving_ the `stderr` writer from the `Pipe` struct, then to be sure we only need a read lock (and you've answered your own question, in a way: the solution would be to change the existing `sync.Mutex to a `sync.RWMutex`).

But the question arises whether it's safe for multiple concurrent `Exec`s to all write to the same writer. We can't know that it is. On the other hand, if we were to enforce serialization with a separate mutex for this error writer, then we'd lose the concurrency.

On balance I think we should let the user worry about that, don't you? After all, the configured stderr is overwhelmingly likely to be either `os.Stderr` or some file, in which case you might end up with interleaved writes looking messy, but no actual crashes.

As far as the `p.mu` goes, I think we can keep it as a write-lock for now: it doesn't seem like this is in the hot path of anything.

mahadzaryab1 replied:
Sounds good! Thank you for the detailed explanation @bitfield

bitfield commented on script_test.go:
```diff
@@ -1850,6 +1850,19 @@ func TestReadReturnsErrorGivenReadErrorOnPipe(t *testing.T) {
 
+// TestWithStdErrAfterExec is a regression test that was added to test against
+// a race condition for Pipe.stderr.
+func TestWithStdErrAfterExec(t *testing.T) {
```
Testing for data races is not an exact science, but it's generally a good idea to put the doubtful code under as much stress as possible—calling it a few hundred times from two or more concurrent goroutines, for example. I don't think you mentioned whether this test failed without your fix, with or without the race detector? I wouldn't have thought it would.

Can you put together a test that fails _fairly_ reliably without the mutex lock? I don't know if you've read my book [Know Go](https://bitfieldconsulting.com/books/generics) (few people have, so I don't blame you), but there's a section in there on "smoke testing" code for concurrency safety. Here's one of the examples:

```go
s := NewSetC(1, 2, 3)
var wg sync.WaitGroup
wg.Add(1)
go func() {
	for i := range 1000 {
		s.Add(i)
	}
	wg.Done()
}()
for i := range 1000 {
	_ = s.All()
}
wg.Wait()
fmt.Println("We made it!")
// Output:
// We made it!
```

mahadzaryab1 replied:
@bitfield Thanks for the detailed comment. Do we want a test that fails without the race detector on? The current test I've written fails with the race detector on but not without.

bitfield replied:
If it fails with the race detector on, that's good enough!

bitfield reviewed (changes requested):
Looking good so far!

bitfield commented:
@mahadzaryab1 I'm happy if you're happy—do you want to go ahead and do a final quality pass, checking documentation comments, alphabetical order, `gotestdox`-friendly test names, and so on? You know the house style now 😄

mahadzaryab1 commented:
> @mahadzaryab1 I'm happy if you're happy—do you want to go ahead and do a final quality pass, checking documentation comments, alphabetical order, `gotestdox`-friendly test names, and so on? You know the house style now 😄

@bitfield Done! Let me know if there's anything I missed.

bitfield reviewed (approved):
Super nice, @mahadzaryab1! Thanks a lot. I'll wait till you're done on #208, and then we can roll all these changes up into a new release.
